examples/elixir_processor/lib/elixir_processor.ex

```diff
@@ -0,0 +1,176 @@
+  #   event: %Exmld.KinesisWorker.Datum{
+  #       opaque: {"us-west-2", "erlang-processor-kinesis-test"},
+  #       shard_id: "shardId-000000000001",
+  #       stream_record: {:stream_record, "12345", 946684800,
+  #                        {:sequence_number, 12345, 0, :undefined, :undefined},
+  #                        " .. record data .. "}},
+  #   stage: #PID<0.136.0>,
+  #   worker: #PID<0.862.0>}
+  #
+  defp flow_extract(%Exmld.KinesisStage.Event{
+         event: %Exmld.KinesisWorker.Datum{stream_record: record},
+         stage: stage,
+         worker: worker
+       }) do
+    case record do
+      # handle a heartbeat.  the second element of the tuple will vary so heartbeats get
+      # distributed among reducers, so the elements must be swapped since we're using the
+      # first element as a partition key.
+      {:heartbeat, x} ->
+        [{x, :heartbeat}]
+
+      # in a real application, sub-records could be extracted from Event here.  if using a
+      # custom non-KPL aggregation scheme, this should associate each sub-record with a
+      # faked sequence number having the same base as the parent record, and appropriate
+      # 'user_sub' (sub-record index) and 'user_total' (total number of extracted
+      # sub-records) fields.  then when later notifying exmld of record disposition, it
+      # can properly track sub-record processing and advance the checkpoint beyond the
+      # parent record if all of its sub-records were processed.
+      #
+      # two records having the same key (first tuple element) here will be handled by the
+      # same reducer.  in general, the key should be consistently derived from some
+      # attribute of the record/item being processed.
+      _ when Record.is_record(record, :stream_record) ->
+        sn =
+          Exmld.stream_record(record, :sequence_number)
+          |> Exmld.sequence_number(user_sub: 0)
+          |> Exmld.sequence_number(user_total: 1)
+
+        item = %Item{
+          value: record,
+          token: %Token{stage: stage, worker: worker, sequence_number: sn}
+        }
+
+        [{:erlang.phash2(item), item}]
+    end
+  end
+
+  # process an item extracted from a record (or a heartbeat).  this occurs in a reducer
+  # whose initial state is given by 'state0' in flow_spec/3 above.  it returns an updated
+  # state after processing the event (and possibly flushing/updating the state
+  # accordingly).  here, we simply add the item to the current batch and possibly flush
+  # the batch.
+  defp flow_process_event({_key, item}, state) do
+    state
+    |> flow_add_record(item)
+    |> maybe_flush()
+  end
+
+  defp flow_add_record(state, :heartbeat) do
+    state
+  end
+
+  defp flow_add_record(%__MODULE__{pending_items: pending} = state, item) do
+    %{state | pending_items: [item | pending]}
+  end
+
+  # possibly process the current pending batch of records if of the appropriate size or
+  # enough time has elapsed:
+  defp maybe_flush(state) do
```

Comment:
Might be worth to use [Flow.Window.trigger* family of functions](https://hexdocs.pm/flow/Flow.Window.html#trigger/3), or at least mention these here (which seems like the way to do this kind of things).

Reply:
Good idea, will look into it (and also making the exmld user interface more convenient for doing that, e.g., via options to `Exmld.flow/6` or `Exmld.from_stages/1`).

Reply:
@marcelog: Added use of a time-based window for the elixir example (& added a new option accepted by `Exmld.flow/6` supporting the new example).